trinity/protocol/common/trackers.py

```diff
@@ -0,0 +1,199 @@
+        """
+        if not self.total_msgs:
+            return 'None'
+        avg_rtt = self.total_response_time / self.total_msgs
+        if not self.total_items:
+            items_per_second = 0.0
+        else:
+            items_per_second = self.total_response_time / self.total_items
+
+        # msgs: total number of messages
+        # items: total number of items
+        # rtt: round-trip-time (avg/ema)
+        # ips: items-per-second (avg/ema)
+        # timeouts: total number of timeouts
+        # missing: total number of missing response items
+        # quality: 0-100 for how complete responses are
+        return (
+            'msgs=%d  items=%d  rtt=%.2f/%.2f  ips=%.5f/%.5f  '
+            'timeouts=%d  missing=%d  quality=%d'
+        ) % (
+            self.total_msgs,
+            self.total_items,
+            avg_rtt,
+            self.round_trip_time_ema.value,
+            items_per_second,
+            self.items_per_second_ema.value,
+            self.total_timeouts,
+            self.total_missing,
+            int(self.response_quality_ema.value),
+        )
+
+    def record_timeout(self) -> None:
+        self.total_msgs += 1
+        self.total_timeouts += 1
+        self.response_quality_ema.update(0)
+
+    def record_response(self,
+                        time: float,
+                        request: TRequest,
+                        result: TResult) -> None:
+        self.total_msgs += 1
+
+        request_size = self._get_request_size(request)
+        response_size = self._get_result_size(result)
+        num_items = self._get_result_item_count(result)
+
+        if request_size is None:
+            pass
+        elif response_size > request_size:
+            self.logger.warning(
+                "%s got oversized response.  requested: %d  received: %d",
+                type(self).__name__,
+                request_size,
+                response_size,
+            )
+        else:
+            self.total_missing += (request_size - response_size)
+            if request_size == 0:
+                self.logger.warning(
+                    "%s encountered request for zero items. This should never happen",
+                    type(self).__name__,
+                )
+                self.response_quality_ema.update(100)
+            elif response_size == 0:
+                self.response_quality_ema.update(0)
+            else:
+                percent_returned = 100 * response_size / request_size
+                self.response_quality_ema.update(percent_returned)
+
+        self.total_items += num_items
+        self.total_response_time += time
+        self.round_trip_time_ema.update(time)
+
+        if time > 0:
+            throughput = num_items / time
+            self.items_per_second_ema.update(throughput)
+        else:
+            self.logger.warning(
+                "%s encountered response time of zero.  This should never happen",
```

Comment:
`<= zero`

I wouldn't be surprised at all if something like this happened because someone measured duration using `time.time()` instead of `time.perf_counter()` (and then daylight savings happens).